File: go/src/2018/day12/day12.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"utils"
)

func parsePotsAndPatterns(filepath string) (pots string, rules []string) {
	lines := utils.ReadFileToLines(filepath)
	initialStatePattern := "initial state: (.*)"
	initialStateRegexp := regexp.MustCompile(initialStatePattern)
	rulePattern := "(.*) => #"
	ruleRegexp := regexp.MustCompile(rulePattern)
	for _, line := range lines {
		initialStateMatch := initialStateRegexp.FindStringSubmatch(line)
		if initialStateMatch != nil {
			pots = initialStateMatch[1]
		}
		ruleMatch := ruleRegexp.FindStringSubmatch(line)
		if ruleMatch != nil {
			rules = append(rules, ruleMatch[1])
		}
	}
	return
}

func applyRules(input string, rules []string) string {
	output := strings.Repeat(".", len(input))
	for _, r := range rules {
		ruleRegex := regexp.MustCompile(regexp.QuoteMeta(r))
		i := 0
		for ruleMatch := ruleRegex.FindStringIndex(input); ruleMatch != nil; ruleMatch = ruleRegex.FindStringIndex(input[i:]) {
			i += ruleMatch[0] + 1
			output = output[:i+1] + "#" + output[i+1+1:]
		}
	}
	return output
}

func sumPots(pots string, offset int) (sum int) {
	for i, p := range pots {
		if string(p) == "#" {
			sum += i - offset
		}
	}
	return
}

func main() {
	pots, rules := parsePotsAndPatterns(os.Args[1])
	offset := 10
	pots = strings.Repeat(".", offset) + pots + strings.Repeat(".", 150)
	for g := 0; ; g++ {
		tempPots := applyRules(pots, rules)

		if g == 19 {
			fmt.Printf("sum #20: %v\n", sumPots(tempPots, offset-g))
		}

		if pots == tempPots[1:]+"." {
			break // Pattern is now re-occuring but with an offset
		}

		pots = tempPots
		offset++
	}
	fmt.Printf("sum #50000000000: %v\n", sumPots(pots, offset-5000000000))
}

File: python/2018/day12.py
import re


def parsePotsAndPatters(filepath):
    f = open(filepath, "r")
    lines = f.read().splitlines()
    initialStateRegexp = r"initial state: (.*)"
    initialStatePattern = re.compile(initialStateRegexp)
    initialStateMatch = initialStatePattern.search(lines[0])
    pots = initialStateMatch.group(1)
    ruleRegexp = r"(.*) => #"
    rulePattern = re.compile(ruleRegexp)
    rules = []
    for line in lines[1:]:
        ruleMatch = rulePattern.search(line)
        if ruleMatch is None:
            continue
        rules.append(ruleMatch.group(1))
    return (pots, rules)


def incrementGeneration(pots, rules):
    newPots = "." * len(pots)
    for r in rules:
        pattern = re.compile("(?=" + re.escape(r) + ")")
        for match in pattern.finditer(pots):
            newPots = newPots[: match.start() + 2] + "#" + newPots[match.start() + 3 :]
    return newPots


def sumPotIndices(pots, offset):
    sum = 0
    for i in range(len(pots)):
        if pots[i] == "#":
            sum += i - offset
    return sum


def main():
    pots, rules = parsePotsAndPatters("day12/input")
    startOffset = 10
    pots = "." * 10 + pots + "." * 150
    g = 0
    while True:
        newPots = incrementGeneration(pots, rules)
        if g == 19:
            print("sum #20:", sumPotIndices(newPots, startOffset - g))

        if newPots[1:] + "." == pots:
            print(
                "sum #50000000000:",
                sumPotIndices(newPots, startOffset + 1 - 5000000000),
            )
            break

        pots = newPots
        g += 1
        startOffset += 1


if __name__ == "__main__":
    main()
